driver-core/src/test/functional/com/mongodb/internal/connection/AwsAuthenticationSpecification.groovy

```diff
@@ -0,0 +1,85 @@
+package com.mongodb.internal.connection
+
+import com.mongodb.ClusterFixture
+import com.mongodb.MongoCommandException
+import com.mongodb.MongoCredential
+import com.mongodb.ServerAddress
+import com.mongodb.async.FutureResultCallback
+import com.mongodb.connection.ClusterId
+import com.mongodb.connection.ServerId
+import com.mongodb.connection.SocketSettings
+import com.mongodb.connection.SocketStreamFactory
+import com.mongodb.connection.netty.NettyStreamFactory
+import org.bson.BsonDocument
+import org.bson.BsonString
+import spock.lang.IgnoreIf
+import spock.lang.Specification
+
+import static com.mongodb.AuthenticationMechanism.MONGODB_AWS
+import static com.mongodb.ClusterFixture.getConnectionString
+import static com.mongodb.ClusterFixture.getCredential
+import static com.mongodb.ClusterFixture.getSslSettings
+import static com.mongodb.internal.connection.CommandHelper.executeCommand
+import static java.util.concurrent.TimeUnit.SECONDS
+
+@IgnoreIf({ getCredential() == null || getCredential().getAuthenticationMechanism() != MONGODB_AWS })
+class AwsAuthenticationSpecification extends Specification {
+
+    def 'should not authorize when not authenticated'() {
+        given:
+        def connection = createConnection(async, null)
+
+        when:
+        openConnection(connection, async)
+        executeCommand(getConnectionString().getDatabase(), new BsonDocument('count', new BsonString('test')), connection)
+
+        then:
+        thrown(MongoCommandException)
+
+        cleanup:
+        connection?.close()
+
+        where:
+        async << [true, false]
+    }
+
+    def 'should authorize when successfully authenticated'() {
+        given:
+        def connection = createConnection(async, getCredential())
+
+        when:
+        openConnection(connection, async)
+        executeCommand(getConnectionString().getDatabase(), new BsonDocument('count', new BsonString('test')), connection)
+
+        then:
+        true
+
+        cleanup:
+        connection?.close()
+
+        where:
+        async << [true, false]
+    }
+
+    private static InternalStreamConnection createConnection(final boolean async, final MongoCredential credential) {
+        new InternalStreamConnection(
+                new ServerId(new ClusterId(), new ServerAddress(getConnectionString().getHosts().get(0))),
+                async ? new NettyStreamFactory(SocketSettings.builder().build(), getSslSettings())
```

Comment:
Prefer: `AsynchronousSocketChannelStreamFactory` instead of netty, now that is the main async factory.

Reply:
Done: